src/main/java/com/google/api/codegen/config/GrpcGapicRetryMapping.java

```diff
@@ -0,0 +1,163 @@
+      paramsDefMap.putIfAbsent(paramsBuilder.getName(), paramsBuilder.build());
+
+      // construct retry codes from RetryPolicy
+      RetryCodesDefinitionProto.Builder codesBuilder = RetryCodesDefinitionProto.newBuilder();
+      codesBuilder.setName(policyName + "_codes");
+      retryPolicy
+          .getRetryableStatusCodesList()
+          .forEach(code -> codesBuilder.addRetryCodes(code.name()));
+      codesDefMap.putIfAbsent(codesBuilder.getName(), codesBuilder.build());
+
+      // apply specific method config (overwrites) or apply service config to all methods (does
+      // not overwrite method-specific policies)
+      for (Name name : methodConfig.getNameList()) {
+        String serviceBase = name.getService();
+
+        if (!Strings.isNullOrEmpty(name.getMethod())) {
+          String fullName = fullyQualifiedName(serviceBase, name.getMethod());
+          methodCodesMap.put(fullName, codesBuilder.getName());
+          methodParamsMap.put(fullName, paramsBuilder.getName());
+          continue;
+        }
+
+        Interface interProto = protoInterfaces.get(serviceBase);
+        for (Method method : interProto.getMethods()) {
+          String fullName = method.getFullName();
+          methodCodesMap.putIfAbsent(fullName, codesBuilder.getName());
+          methodParamsMap.putIfAbsent(fullName, paramsBuilder.getName());
+        }
+      }
+    }
+
+    // make immutable
+    ImmutableMap<String, String> codes = ImmutableMap.copyOf(methodCodesMap);
+    ImmutableMap<String, String> params = ImmutableMap.copyOf(methodParamsMap);
+    ImmutableMap<String, RetryCodesDefinitionProto> codesProto = ImmutableMap.copyOf(codesDefMap);
+    ImmutableMap<String, RetryParamsDefinitionProto> paramsProto =
+        ImmutableMap.copyOf(paramsDefMap);
+
+    return new AutoValue_GrpcGapicRetryMapping.Builder()
+        .setMethodCodesMap(codes)
+        .setMethodParamsMap(params)
+        .setCodesDefMap(codesProto)
+        .setParamsDefMap(paramsProto)
+        .build();
+  }
+
+  public abstract ImmutableMap<String, String> methodCodesMap();
+
+  public abstract ImmutableMap<String, String> methodParamsMap();
+
+  public abstract ImmutableMap<String, RetryCodesDefinitionProto> codesDefMap();
+
+  public abstract ImmutableMap<String, RetryParamsDefinitionProto> paramsDefMap();
+
+  public static Builder builder() {
+    return new AutoValue_GrpcGapicRetryMapping.Builder();
+  }
+
+  /**
+   * Converts a float to a double rounding to the nearest hundredth.
+   *
+   * <p>This is necessary because when a float is converted to a double the double is more precise.
+   * When that same double is printed (e.g. in a template), the more precise value is used. For
+   * example, a float of 1.3 is printed as 1.2999999523162842. This is undesirable because this
+   * multiplier does not need to be this precise (it was originally a float) and it is not the
+   * expected value.
+   */
+  private static double floatToDouble(float f) {
```

Comment:
Your understanding of my hack here is entirely correct! And your explanation/research is greatly appreciated.

I will try to figure out how we can do the formatting prior to printing.

Reply:
OK, so this is actually the only place the value is assigned before being printed (it's just passed around between here and the snippets). The snippets call `{@def.params.getRetryDelayMultiplier}` which is a method on the `RetryParamsDefinitionProto`. I could add some method to the snippets and then change all of the snippets to use that for formatting the value from the definition? This seems overly complicated...

Or I could do something like this, but I guess it's still changing the value too early like you said not to do...
```java
private static double floatToDouble(float f) {
    String formatted = new DecimalFormat(".##").format((double) f);
    return Double.parseDouble(formatted);
}
```

WDYT?

Reply:
what about consistently using floats/doubles everywhere?

Reply:
These values represent fields from protos that are defined as `float` (gRPC ServiceConfig) and `double` (GAPIC Config). So we _could_ change one of them, it would really have to be the GAPIC Config to not diverge from gRPC, but that would result in cascading changes elsewhere in the code base. I'm not sure it's worth the effort :smile:

Reply:
Can we use 1.25 instead of 1.3? :)

Reply:
> Can we use 1.25 instead of 1.3? :)

No, because the value is actually `1.3` in the config in this case, but conversion to `double` does the above mangling^^

Reply:
**tl;dr;** Some brief history of gapic-generator architecture is below. I helped me, when I finally understood what was going on in gapic-generator, and is kind of relevant here as we decide the proper place where the toString conversion should go.

**tl; dr; 2** If you think it is more complicated than it should be, then you are not alone... I believe that the "thing" described below is the main source of gapic-generator complications and why people don't like it. Fixing that should be doable though (different topic).

**The very biref history of gapic-generator's architecture**
On a very high level the architecture of the gapic-generator tries to follow MVC (Model-View-Controller) model (if it applies well here and if it was a good idea is a different story...):
1. Consume all config values from everywhere (protos, yamls, command line args, stuff...). We can think about those as the **"Model"**.
2. The gathered config values (see item 1) are used for two main purposes:
    1. Control generation flow (i.e. determine which files should be generated and how). The **"Controller"** part (well, the term is applied here very loosely, but close enough)
    2. Determine what exactly should be put in those generated files (i.e. the **actual text**).
3. The actual text part is determined by a combination of in `.snip` files and corresponding `*View` objects (the **"View"** part). GAPIC simply takes a snip file, a View object and applies one another to get a text file output.
     1. Notice that the methods of the view classes (like `RetryParamsDefinitionView` in our case) almost solely return either `String`s or collections of `Strings`. This is so because they are all about "text".
     2. Something like `{@def.params.getMaxRetryDelayMillis}` in a snip file literally means a call to `def.params().getMaxRetryDelayMillis()`, where `def` is an object of a `*View` type (`RetryParamsDefinitionView` in our case).
4. To create the `*View` objects, a corresponding `*Transformer` object is used (i guess transformers loosely fall in to the Controller part). It will be the `RetryDefinitionsTransformer` class in our case.

**How it applies to our concrete problem**
When template parameters are applied in snip files, those are taken from `RetryParamsDefinitionView` object. But that object has nested objects in it (`RetryDefinitionsTransformer`), and those are returned directly. Which is **fine if you do not need any special string representation logic** in place for that object. But looks like we have it here (to print the float properly). 

To solve it "idiomatically" for gapic-generator, we need a nested View object to be returned. Nested view objects are very common in gapic-generator (for example `ServiceDocView` has `exampleApiMethod()` which returns a nested view `ApiMethodView`).

So we need to create a class, something like `RetryParamsDefinitionProtoView`, and control there the String representation of its fields.

```java
public class RetryParamsDefinitionProto {
  // For fields which don't have custom logic, just propagate them as is
  public String getRetryDelayMultiplier() {
    return new DecimalFormat(".##").format(proto.getRetryDelyMultiplier());
  }
  // ...
}
```

Then, when `RetryParamsDefinitionView` is created in `RetryDefinitionsTransformer` the view object is put (instead of RetryParamsDefinitionProto) such tat `RetryParamsDefinitionView.params()` returns `RetryParamsDefinitionProtoView` instead of `RetryParamsDefinitionProto` and when called from snip file that will output proper string.

Reply:
**tl;dr;** Some brief history of gapic-generator architecture is below. I helped me, when I finally understood what was going on in gapic-generator, and is kind of relevant here as we decide the proper place where the toString conversion should go.

**tl; dr; 2** If you think it is more complicated than it should be, then you are not alone... I believe that the "thing" described below is the main source of gapic-generator complications and why people don't like it. Fixing that should be doable though (different topic).

**The very biref history of gapic-generator's architecture**
On a very high level the architecture of the gapic-generator tries to follow MVC (Model-View-Controller) model (if it applies well here and if it was a good idea is a different story...):
1. Consume all config values from everywhere (protos, yamls, command line args, stuff...). We can think about those as the **"Model"**.
2. The gathered config values (see item 1) are used for two main purposes:
    1. Control generation flow (i.e. determine which files should be generated and how). The **"Controller"** part (well, the term is applied here very loosely, but close enough)
    2. Determine what exactly should be put in those generated files (i.e. the **actual text**).
3. The actual text part is determined by a combination of in `.snip` files and corresponding `*View` objects (the **"View"** part). GAPIC simply takes a snip file, a View object and applies one another to get a text file output.
     1. Notice that the methods of the view classes (like `RetryParamsDefinitionView` in our case) almost solely return either `String`s or collections of `Strings`. This is so because they are all about "text".
     2. Something like `{@def.params.getMaxRetryDelayMillis}` in a snip file literally means a call to `def.params().getMaxRetryDelayMillis()`, where `def` is an object of a `*View` type (`RetryParamsDefinitionView` in our case).
4. To create the `*View` objects, a corresponding `*Transformer` object is used (i guess transformers loosely fall in to the Controller part). It will be the `RetryDefinitionsTransformer` class in our case.

**How it applies to our concrete problem**
When template parameters are applied in snip files, those are taken from `RetryParamsDefinitionView` object. But that object has nested objects in it (`RetryDefinitionsTransformer`), and those are returned directly. Which is **fine if you do not need any special string representation logic** in place for that object. But looks like we have it here (to print the float properly). 

To solve it "idiomatically" for gapic-generator, we need a nested View object to be returned. Nested view objects are very common in gapic-generator (for example `ServiceDocView` has `exampleApiMethod()` which returns a nested view `ApiMethodView`).

So we need to create a class, something like `RetryParamsDefinitionProtoView`, and control there the String representation of its fields.

```java
public class RetryParamsDefinitionProto {
  // For fields which don't have custom logic, just propagate them as is
  public String getRetryDelayMultiplier() {
    return new DecimalFormat(".##").format(proto.getRetryDelyMultiplier());
  }
  // ...
}
```

Then, when `RetryParamsDefinitionView` is created in `RetryDefinitionsTransformer` the view object is put (instead of RetryParamsDefinitionProto) such tat `RetryParamsDefinitionView.params()` returns `RetryParamsDefinitionProtoView` instead of `RetryParamsDefinitionProto` and when called from snip file that will output proper string.

Reply:
> So we need to create a class, something like RetryParamsDefinitionProtoView, and control there the String representation of its fields.

Just to be clear, is that the recommendation? Or is this just an example of how it would be done "idiomatically" given the history?

Reply:
@noahdietz At this point I'm wondering why RetryPolicy.backoff_multiplier is declared as float (they could just declare it as double and make everybody happy). But if we can't change that, we do need to keep string representation and exact values separate. 

In general, we need to take double values under control. Probably the least verbose way we can do it is changing `RetryParamsDefinitionView` class by adding the `RetryParamsDefinitionProtoView` in it:

```java
@AutoValue
public abstract class RetryParamsDefinitionView {

  // ... Unchanged methods ...
  // Replace returned value from RetryParamsDefinitionProto to RetryParamsDefinitionProtoView
  public abstract RetryParamsDefinitionProtoView params();
  
  // ... Unchanged methods ... 

  // Add the view class as public static internal (to minimize spawning new classes and files)
  public static class RetryParamsDefinitionProtoView {
    private final RetryParamsDefinitionProto params;

    public RetryParamsDefinitionProtoView(RetryParamsDefinitionProto params) {
      this.params = params;
    }

    public String getName() {
      return params.getName();
    }

    public long getInitialRetryDelayMillis() {
      return params.getInitialRetryDelayMillis();
    }

    public String getRetryDelayMultiplier() {
      return new DecimalFormat(".####").format(params.getRetryDelayMultiplier());
    }

    public long getMaxRetryDelayMillis() {
      return params.getMaxRetryDelayMillis();
    }

    public long getInitialRpcTimeoutMillis() {
      return params.getInitialRpcTimeoutMillis();
    }

    public String getRpcTimeoutMultiplier() {
      return new DecimalFormat(".####").format(params.getRpcTimeoutMultiplier());
    }

    public long getMaxRpcTimeoutMillis() {
      return params.getMaxRpcTimeoutMillis();
    }

    public long getTotalTimeoutMillis() {
      return params.getTotalTimeoutMillis();
    }
  }
  
  @AutoValue.Builder
  public abstract static class Builder {
    // ... Unchanged methods in Builder class ...

    // Change input parameter value from RetryParamsDefinitionProto to RetryParamsDefinitionProtoView
    public abstract Builder params(RetryParamsDefinitionProtoView val);
  }
}
```

Then in `RetryDefinitionsTransformer` change the line 127 to:
```java
      view.params(
          new RetryParamsDefinitionView.RetryParamsDefinitionProtoView(retryCodesDef.getValue()));

```

That should be it (no more changes required anywhere, including snip files).

Reply:
Thanks a lot @vam-google, this worked great for Java. I had to tweak the formatting methods to return `""` when the value is `0` because it was printing `.0`. Empty string means the snippet won't print the setter for that value, which is the preferred behavior, vs. printing the setter with `.0` in it.

One issue however is the that this doesn't work with languages that load client config from a JSON file: php, python, nodejs and ruby. This View we've changed never gets called, and instead, the JSON representation of the value is just printed, meaning we get `1.2999999523162842` instead of `1.3`. I'm trying to find the section that handles this case...

Reply:
Correction, a different method is being called in the [clientconfig/common.snip](https://github.com/googleapis/gapic-generator/blob/8de8e350194d1a72ac171a57e368e7a9e7e416a8/src/main/resources/com/google/api/codegen/clientconfig/common.snip#L62) that generates the JSON client config

Reply:
@hzyi-google mentioned it before, `RetryParamsDefinitionProto` is one of the few config classes using the Model directly, rather than an intermediate class (like RetryCodesConfig). This makes it hard to build on for things like this. The concrete model is being passed directly to the snippet.

Reply:
Grrr... Such a mess =(... we have such a ridiculous mess in gapic-generator...

So when I look to it closer, we actually already have:
- RetryCodesDefinitionView
- RetryConfigDefinitionView
- RetryParamsDefinitionView

We already have 3 classes which do same thing... People just add stuff completely disregarding what is already there. Such a mess. The "idiomatic" way would be fixing that mess and introducing one view which covers all cases. 

**For the clientconfig/common.snip:**
That should not even be part of gapic generator (it should not generate those type of files)... So much garbage =(. To process those the `LegacyGapicGenrator` is used  which does not work with views...

**What to do**
Ok, at this point it is just me torturing you and forcing to the yak shaving... Scrap it, it is an overwhelming mess that is not your fault, so expecting you to fix it as part of this pr will be not ok. 

Probably the best thing we can do here without starting the yak shaving thing will be changing this method implementation to something like :
```java
return Double.valueOf(Float.toString(1.3F)).doubleValue()
```
This converts `float` to `double` via their's string representation "without loosing" or explicitly specifying precision. This is fundamentally wrong, but kind of ok for us, because the values themselves come from strings and go to strings (the gRPC service config, the source of truth of that value is actually a string).

Sorry, I did not realize that retries have many "views" (this must be fixed) and that the `LegacyGapicGenerator` is still in use.

Reply:
Thanks for all of your help, @vam-google. I went with your final suggestion here.

@AlanGasperini what Vadym mentions might be nice technical debt to target in a fix-it week whenever one of those things comes around again.

Reply:
sgtm